File: windows/PokeFolio.Desktop/Backend/PokeFolioApiClient.cs
using System.Net;
using System.Security.Cryptography;
using PokeFolio.Desktop.Security;

namespace PokeFolio.Desktop.Backend;

public sealed class PokeFolioApiClient : IDisposable
{
    private const int MaximumAuthResponseBytes = 128 * 1024;
    private const int MaximumSyncResponseBytes = 8 * 1024 * 1024;

    private readonly IRefreshTokenStore refreshTokenStore;
    private readonly PokeFolioApiTransport transport;
    private readonly SemaphoreSlim sessionGate = new(1, 1);
    private SessionState? session;
    private int disposed;

    public PokeFolioApiClient(Uri backendOrigin, IRefreshTokenStore refreshTokenStore)
    {
        this.refreshTokenStore = refreshTokenStore ??
            throw new ArgumentNullException(nameof(refreshTokenStore));
        transport = new PokeFolioApiTransport(backendOrigin);
    }

    internal PokeFolioApiClient(
        Uri backendOrigin,
        IRefreshTokenStore refreshTokenStore,
        HttpMessageHandler handler)
    {
        this.refreshTokenStore = refreshTokenStore ??
            throw new ArgumentNullException(nameof(refreshTokenStore));
        transport = new PokeFolioApiTransport(backendOrigin, handler);
    }

    public Uri BackendOrigin => transport.BackendOrigin;

    public PokeFolioSession? CurrentSession => ToPublicSession(Volatile.Read(ref session));

    public static bool IsAllowedBackendOrigin(Uri? origin) =>
        PokeFolioApiTransport.IsAllowedBackendOrigin(origin);

    public async Task<PokeFolioAuthenticationResult> RegisterAsync(
        string email,
        string password,
        string deviceName,
        CancellationToken cancellationToken = default) =>
        await StartSessionAsync(
            "/api/v1/auth/register",
            email,
            password,
            deviceName,
            cancellationToken);

    public async Task<PokeFolioAuthenticationResult> LoginAsync(
        string email,
        string password,
        string deviceName,
        CancellationToken cancellationToken = default) =>
        await StartSessionAsync(
            "/api/v1/auth/login",
            email,
            password,
            deviceName,
            cancellationToken);

    public async Task<PokeFolioAuthenticationResult> RestoreSessionAsync(
        CancellationToken cancellationToken = default)
    {
        ThrowIfDisposed();
        await sessionGate.WaitAsync(cancellationToken);
        try
        {
            SessionState? current = Volatile.Read(ref session);
            if (current is not null)
            {
                return PokeFolioAuthenticationResult.Success(ToPublicSession(current)!);
            }

            RefreshTokenCredential? credential = refreshTokenStore.Load();
            return credential is null
                ? PokeFolioAuthenticationResult.Failed(SessionMissingProblem())
                : await RefreshCoreAsync(credential, cancellationToken);
        }
        finally
        {
            sessionGate.Release();
        }
    }

    public async Task<PokeFolioApiResponse> PushSyncOperationsAsync(
        string operationBatchJson,
        CancellationToken cancellationToken = default)
    {
        byte[] body = PokeFolioApiPayloads.ValidateSyncOperationBatch(operationBatchJson);
        try
        {
            return await SendAuthenticatedAsync(
                HttpMethod.Post,
                "/api/v1/sync/operations",
                body,
                cancellationToken);
        }
        finally
        {
            CryptographicOperations.ZeroMemory(body);
        }
    }

    public async Task<PokeFolioApiResponse> PullSyncChangesAsync(
        string? cursor = null,
        int limit = 100,
        CancellationToken cancellationToken = default)
    {
        if (limit is < 1 or > 500)
        {
            throw new ArgumentOutOfRangeException(nameof(limit), "Sync page size must be 1 to 500.");
        }
        if (cursor?.Length > 2048)
        {
            throw new ArgumentException("Sync cursor exceeds 2048 characters.", nameof(cursor));
        }

        string path = $"/api/v1/sync/changes?limit={limit}";
        if (!string.IsNullOrEmpty(cursor))
        {
            path += "&cursor=" + Uri.EscapeDataString(cursor);
        }
        return await SendAuthenticatedAsync(HttpMethod.Get, path, body: null, cancellationToken);
    }

    public async Task<PokeFolioLogoutResult> LogoutAsync(
        CancellationToken cancellationToken = default)
    {
        ThrowIfDisposed();
        await sessionGate.WaitAsync(cancellationToken);
        PokeFolioRawResponse? response = null;
        try
        {
            SessionState? current = Volatile.Read(ref session);
            if (current is null)
            {
                RefreshTokenCredential? stored = refreshTokenStore.Load();
                if (stored is not null)
                {
                    PokeFolioAuthenticationResult restored =
                        await RefreshCoreAsync(stored, cancellationToken);
                    current = restored.Succeeded ? Volatile.Read(ref session) : null;
                    if (current is null)
                    {
                        return new PokeFolioLogoutResult(false, restored.Problem);
                    }
                }
            }

            if (current is null) return new PokeFolioLogoutResult(false);
            response = await transport.SendAsync(
                HttpMethod.Post,
                "/api/v1/auth/logout",
                body: null,
                current.AccessToken,
                MaximumAuthResponseBytes,
                cancellationToken);
            if (response.Status == (int)HttpStatusCode.Unauthorized)
            {
                RefreshTokenCredential? stored = refreshTokenStore.Load();
                if (stored is not null)
                {
                    PokeFolioAuthenticationResult refreshed =
                        await RefreshCoreAsync(stored, cancellationToken);
                    current = refreshed.Succeeded ? Volatile.Read(ref session) : null;
                    if (current is not null)
                    {
                        CryptographicOperations.ZeroMemory(response.Body);
                        response = await transport.SendAsync(
                            HttpMethod.Post,
                            "/api/v1/auth/logout",
                            body: null,
                            current.AccessToken,
                            MaximumAuthResponseBytes,
                            cancellationToken);
                    }
                }
            }

            return response.IsSuccess
                ? new PokeFolioLogoutResult(true)
                : new PokeFolioLogoutResult(
                    false,
                    PokeFolioApiPayloads.ParseProblem(response));
        }
        finally
        {
            if (response is not null) CryptographicOperations.ZeroMemory(response.Body);
            Volatile.Write(ref session, null);
            try
            {
                refreshTokenStore.Delete();
            }
            finally
            {
                sessionGate.Release();
            }
        }
    }

    private async Task<PokeFolioAuthenticationResult> StartSessionAsync(
        string path,
        string email,
        string password,
        string deviceName,
        CancellationToken cancellationToken)
    {
        ThrowIfDisposed();
        byte[] body = PokeFolioApiPayloads.SerializeLogin(email, password, deviceName);

        await sessionGate.WaitAsync(cancellationToken);
        PokeFolioRawResponse? response = null;
        try
        {
            response = await transport.SendAsync(
                HttpMethod.Post,
                path,
                body,
                accessToken: null,
                MaximumAuthResponseBytes,
                cancellationToken);
            if (!response.IsSuccess)
            {
                return PokeFolioAuthenticationResult.Failed(
                    PokeFolioApiPayloads.ParseProblem(response));
            }

            PokeFolioAuthSessionEnvelope envelope =
                PokeFolioApiPayloads.ParseSession(response.Body);
            ActivateSession(envelope);
            return PokeFolioAuthenticationResult.Success(ToPublicSession(session)!);
        }
        finally
        {
            if (response is not null) CryptographicOperations.ZeroMemory(response.Body);
            CryptographicOperations.ZeroMemory(body);
            sessionGate.Release();
        }
    }

    private async Task<PokeFolioAuthenticationResult> RefreshCoreAsync(
        RefreshTokenCredential credential,
        CancellationToken cancellationToken)
    {
        byte[] body = PokeFolioApiPayloads.SerializeRefresh(credential.RefreshToken);
        PokeFolioRawResponse? response = null;
        try
        {
            response = await transport.SendAsync(
                HttpMethod.Post,
                "/api/v1/auth/refresh",
                body,
                accessToken: null,
                MaximumAuthResponseBytes,
                cancellationToken);
            if (!response.IsSuccess)
            {
                PokeFolioApiProblem problem = PokeFolioApiPayloads.ParseProblem(response);
                if (response.Status is (int)HttpStatusCode.BadRequest or (int)HttpStatusCode.Unauthorized)
                {
                    InvalidateLocalSession();
                }
                return PokeFolioAuthenticationResult.Failed(problem);
            }

            try
            {
                PokeFolioAuthSessionEnvelope envelope =
                    PokeFolioApiPayloads.ParseSession(response.Body);
                if (envelope.Device.Id != credential.DeviceId)
                {
                    throw new InvalidDataException(
                        "Refresh response belongs to a different device session.");
                }
                ActivateSession(envelope);
                return PokeFolioAuthenticationResult.Success(ToPublicSession(session)!);
            }
            catch
            {
                InvalidateLocalSession();
                throw;
            }
        }
        finally
        {
            if (response is not null) CryptographicOperations.ZeroMemory(response.Body);
            CryptographicOperations.ZeroMemory(body);
        }
    }

    private async Task<PokeFolioApiResponse> SendAuthenticatedAsync(
        HttpMethod method,
        string path,
        byte[]? body,
        CancellationToken cancellationToken)
    {
        ThrowIfDisposed();
        SessionResolution resolution = await EnsureSessionAsync(cancellationToken);
        SessionState? current = resolution.Session;
        if (current is null)
        {
            return new PokeFolioApiResponse(
                resolution.Problem!.Status,
                "",
                resolution.Problem);
        }

        PokeFolioRawResponse response = await transport.SendAsync(
            method,
            path,
            body,
            current.AccessToken,
            MaximumSyncResponseBytes,
            cancellationToken);
        if (response.Status == (int)HttpStatusCode.Unauthorized)
        {
            CryptographicOperations.ZeroMemory(response.Body);
            SessionResolution refreshed = await RefreshAfterUnauthorizedAsync(
                current.AccessToken,
                cancellationToken);
            if (refreshed.Session is null)
            {
                return new PokeFolioApiResponse(
                    refreshed.Problem!.Status,
                    "",
                    refreshed.Problem);
            }

            current = refreshed.Session;
            response = await transport.SendAsync(
                method,
                path,
                body,
                current.AccessToken,
                MaximumSyncResponseBytes,
                cancellationToken);
        }

        try
        {
            string responseBody = PokeFolioApiPayloads.DecodeBody(response.Body);
            return response.IsSuccess
                ? new PokeFolioApiResponse(response.Status, responseBody)
                : new PokeFolioApiResponse(
                    response.Status,
                    responseBody,
                    PokeFolioApiPayloads.ParseProblem(response));
        }
        finally
        {
            CryptographicOperations.ZeroMemory(response.Body);
        }
    }

    private async Task<SessionResolution> EnsureSessionAsync(CancellationToken cancellationToken)
    {
        SessionState? current = Volatile.Read(ref session);
        if (current is not null) return new SessionResolution(current, Problem: null);

        await sessionGate.WaitAsync(cancellationToken);
        try
        {
            current = Volatile.Read(ref session);
            if (current is not null) return new SessionResolution(current, Problem: null);
            RefreshTokenCredential? credential = refreshTokenStore.Load();
            if (credential is null)
            {
                return new SessionResolution(Session: null, SessionMissingProblem());
            }
            PokeFolioAuthenticationResult result =
                await RefreshCoreAsync(credential, cancellationToken);
            return result.Succeeded
                ? new SessionResolution(Volatile.Read(ref session), Problem: null)
                : new SessionResolution(Session: null, result.Problem);
        }
        finally
        {
            sessionGate.Release();
        }
    }

    private async Task<SessionResolution> RefreshAfterUnauthorizedAsync(
        string rejectedAccessToken,
        CancellationToken cancellationToken)
    {
        await sessionGate.WaitAsync(cancellationToken);
        try
        {
            SessionState? current = Volatile.Read(ref session);
            if (current is not null &&
                !string.Equals(current.AccessToken, rejectedAccessToken, StringComparison.Ordinal))
            {
                return new SessionResolution(current, Problem: null);
            }

            RefreshTokenCredential? credential = refreshTokenStore.Load();
            if (credential is null)
            {
                return new SessionResolution(Session: null, SessionMissingProblem());
            }
            PokeFolioAuthenticationResult result =
                await RefreshCoreAsync(credential, cancellationToken);
            return result.Succeeded
                ? new SessionResolution(Volatile.Read(ref session), Problem: null)
                : new SessionResolution(Session: null, result.Problem);
        }
        finally
        {
            sessionGate.Release();
        }
    }

    private void ActivateSession(PokeFolioAuthSessionEnvelope envelope)
    {
        var credential = new RefreshTokenCredential(envelope.Device.Id, envelope.RefreshToken);
        try
        {
            refreshTokenStore.Save(credential);
        }
        catch (Exception saveError)
        {
            Volatile.Write(ref session, null);
            try
            {
                refreshTokenStore.Delete();
            }
            catch (Exception deleteError)
            {
                throw new AggregateException(
                    "Rotated refresh token could not be persisted or invalidated.",
                    saveError,
                    deleteError);
            }
            throw new InvalidDataException("Rotated refresh token could not be persisted.", saveError);
        }

        Volatile.Write(ref session, new SessionState(
            envelope.AccessToken,
            envelope.AccessTokenExpiresAt,
            new PokeFolioDevice(
                envelope.Device.Id,
                envelope.Device.Name,
                envelope.Device.Platform,
                envelope.Device.CreatedAt,
                envelope.Device.LastSeenAt)));
    }

    private void InvalidateLocalSession()
    {
        Volatile.Write(ref session, null);
        refreshTokenStore.Delete();
    }

    private static PokeFolioSession? ToPublicSession(SessionState? state) =>
        state is null ? null : new PokeFolioSession(state.Device, state.AccessTokenExpiresAt);

    private static PokeFolioApiProblem SessionMissingProblem() => new(
        (int)HttpStatusCode.Unauthorized,
        "session_missing",
        "No persistent PokeFolio session is available.");

    private void ThrowIfDisposed()
    {
        ObjectDisposedException.ThrowIf(Volatile.Read(ref disposed) != 0, this);
    }

    public void Dispose()
    {
        if (Interlocked.Exchange(ref disposed, 1) != 0) return;
        Volatile.Write(ref session, null);
        transport.Dispose();
    }

    private sealed record SessionState(
        string AccessToken,
        DateTimeOffset AccessTokenExpiresAt,
        PokeFolioDevice Device);

    private sealed record SessionResolution(
        SessionState? Session,
        PokeFolioApiProblem? Problem);
}

File: app/src/main/java/de/pokefolio/app/backend/PokeFolioApiClient.java
package de.pokefolio.app.backend;

import de.pokefolio.app.security.RefreshTokenCredential;
import de.pokefolio.app.security.RefreshTokenStore;

import java.io.Closeable;
import java.io.IOException;
import java.util.Arrays;

/** Thread-safe native account client with one rotating device session. */
public final class PokeFolioApiClient implements Closeable {
    private static final int MAXIMUM_AUTH_RESPONSE_BYTES = 128 * 1024;
    private static final PokeFolioApiProblem SESSION_MISSING = new PokeFolioApiProblem(
            401,
            "session_missing",
            "No persistent PokeFolio session is available.",
            null);

    private final Object sessionGate = new Object();
    private final RefreshTokenStore refreshTokenStore;
    private final PokeFolioApiTransport transport;
    private volatile SessionState session;
    private volatile boolean closed;

    public PokeFolioApiClient(
            PokeFolioBackendConfiguration configuration,
            RefreshTokenStore refreshTokenStore
    ) {
        if (configuration == null || !configuration.isConfigured()) {
            throw new IllegalArgumentException("A valid backend configuration is required.");
        }
        if (refreshTokenStore == null) {
            throw new IllegalArgumentException("Refresh-token storage is required.");
        }
        this.refreshTokenStore = refreshTokenStore;
        this.transport = new PokeFolioHttpTransport(configuration.getBackendOrigin());
    }

    PokeFolioApiClient(
            RefreshTokenStore refreshTokenStore,
            PokeFolioApiTransport transport
    ) {
        if (refreshTokenStore == null || transport == null) {
            throw new IllegalArgumentException("Account client dependencies are required.");
        }
        this.refreshTokenStore = refreshTokenStore;
        this.transport = transport;
    }

    public PokeFolioSession getCurrentSession() {
        return toPublicSession(session);
    }

    public PokeFolioAuthenticationResult register(
            String email,
            String password,
            String deviceName
    ) throws IOException {
        return startSession("/api/v1/auth/register", email, password, deviceName);
    }

    public PokeFolioAuthenticationResult login(
            String email,
            String password,
            String deviceName
    ) throws IOException {
        return startSession("/api/v1/auth/login", email, password, deviceName);
    }

    public PokeFolioAuthenticationResult restoreSession() throws IOException {
        throwIfClosed();
        synchronized (sessionGate) {
            throwIfClosed();
            if (session != null) {
                return PokeFolioAuthenticationResult.success(toPublicSession(session));
            }
            RefreshTokenCredential credential = refreshTokenStore.load();
            return credential == null
                    ? PokeFolioAuthenticationResult.failed(SESSION_MISSING)
                    : refreshCore(credential);
        }
    }

    public PokeFolioLogoutResult logout() throws IOException {
        throwIfClosed();
        synchronized (sessionGate) {
            throwIfClosed();
            PokeFolioRawResponse response = null;
            try {
                SessionState current = session;
                if (current == null) {
                    RefreshTokenCredential stored = refreshTokenStore.load();
                    if (stored != null) {
                        PokeFolioAuthenticationResult restored = refreshCore(stored);
                        current = restored.isSucceeded() ? session : null;
                        if (current == null) {
                            return new PokeFolioLogoutResult(false, restored.getProblem());
                        }
                    }
                }
                if (current == null) return new PokeFolioLogoutResult(false, null);

                response = transport.send(
                        "POST",
                        "/api/v1/auth/logout",
                        null,
                        current.accessToken,
                        MAXIMUM_AUTH_RESPONSE_BYTES);
                if (response.status == 401) {
                    clear(response.body);
                    response = null;
                    RefreshTokenCredential stored = refreshTokenStore.load();
                    if (stored != null) {
                        PokeFolioAuthenticationResult refreshed = refreshCore(stored);
                        current = refreshed.isSucceeded() ? session : null;
                        if (current != null) {
                            response = transport.send(
                                    "POST",
                                    "/api/v1/auth/logout",
                                    null,
                                    current.accessToken,
                                    MAXIMUM_AUTH_RESPONSE_BYTES);
                        }
                    }
                }
                if (response == null) {
                    return new PokeFolioLogoutResult(false, SESSION_MISSING);
                }
                return response.isSuccess()
                        ? new PokeFolioLogoutResult(true, null)
                        : new PokeFolioLogoutResult(
                                false,
                                PokeFolioApiPayloads.parseProblem(response.status, response.body));
            } finally {
                if (response != null) clear(response.body);
                session = null;
                refreshTokenStore.delete();
            }
        }
    }

    PokeFolioApiResponse executeAuthenticated(
            String method,
            String path,
            byte[] body,
            int maximumResponseBytes
    ) throws IOException {
        throwIfClosed();
        SessionResolution resolution = ensureSession();
        SessionState current = resolution.session;
        if (current == null) {
            return new PokeFolioApiResponse(
                    resolution.problem.getStatus(),
                    "",
                    resolution.problem);
        }

        PokeFolioRawResponse response = transport.send(
                method,
                path,
                body,
                current.accessToken,
                maximumResponseBytes);
        if (response.status == 401) {
            clear(response.body);
            SessionResolution refreshed = refreshAfterUnauthorized(current.accessToken);
            if (refreshed.session == null) {
                return new PokeFolioApiResponse(
                        refreshed.problem.getStatus(),
                        "",
                        refreshed.problem);
            }
            response = transport.send(
                    method,
                    path,
                    body,
                    refreshed.session.accessToken,
                    maximumResponseBytes);
        }

        try {
            String responseBody = PokeFolioApiPayloads.decodeUtf8(response.body);
            return response.isSuccess()
                    ? new PokeFolioApiResponse(response.status, responseBody, null)
                    : new PokeFolioApiResponse(
                            response.status,
                            responseBody,
                            PokeFolioApiPayloads.parseProblem(response.status, response.body));
        } finally {
            clear(response.body);
        }
    }

    private PokeFolioAuthenticationResult startSession(
            String path,
            String email,
            String password,
            String deviceName
    ) throws IOException {
        throwIfClosed();
        byte[] body = PokeFolioApiPayloads.serializeLogin(email, password, deviceName);
        synchronized (sessionGate) {
            PokeFolioRawResponse response = null;
            try {
                throwIfClosed();
                response = transport.send(
                        "POST",
                        path,
                        body,
                        null,
                        MAXIMUM_AUTH_RESPONSE_BYTES);
                if (!response.isSuccess()) {
                    return PokeFolioAuthenticationResult.failed(
                            PokeFolioApiPayloads.parseProblem(response.status, response.body));
                }
                PokeFolioApiPayloads.AuthEnvelope envelope =
                        PokeFolioApiPayloads.parseSession(response.body);
                activateSession(envelope);
                return PokeFolioAuthenticationResult.success(toPublicSession(session));
            } finally {
                if (response != null) clear(response.body);
                clear(body);
            }
        }
    }

    private PokeFolioAuthenticationResult refreshCore(
            RefreshTokenCredential credential
    ) throws IOException {
        byte[] body = PokeFolioApiPayloads.serializeRefresh(credential.getRefreshToken());
        PokeFolioRawResponse response = null;
        try {
            response = transport.send(
                    "POST",
                    "/api/v1/auth/refresh",
                    body,
                    null,
                    MAXIMUM_AUTH_RESPONSE_BYTES);
            if (!response.isSuccess()) {
                PokeFolioApiProblem problem =
                        PokeFolioApiPayloads.parseProblem(response.status, response.body);
                if (response.status == 400 || response.status == 401) invalidateLocalSession();
                return PokeFolioAuthenticationResult.failed(problem);
            }
            try {
                PokeFolioApiPayloads.AuthEnvelope envelope =
                        PokeFolioApiPayloads.parseSession(response.body);
                if (!envelope.device.id.equals(credential.getDeviceId())) {
                    throw new IOException("Refresh response belongs to a different device session.");
                }
                activateSession(envelope);
                return PokeFolioAuthenticationResult.success(toPublicSession(session));
            } catch (IOException | RuntimeException error) {
                session = null;
                try {
                    refreshTokenStore.delete();
                } catch (IOException | RuntimeException deleteError) {
                    error.addSuppressed(deleteError);
                }
                throw error;
            }
        } finally {
            if (response != null) clear(response.body);
            clear(body);
        }
    }

    private SessionResolution ensureSession() throws IOException {
        SessionState current = session;
        if (current != null) return new SessionResolution(current, null);
        synchronized (sessionGate) {
            throwIfClosed();
            current = session;
            if (current != null) return new SessionResolution(current, null);
            RefreshTokenCredential credential = refreshTokenStore.load();
            if (credential == null) return new SessionResolution(null, SESSION_MISSING);
            PokeFolioAuthenticationResult result = refreshCore(credential);
            return result.isSucceeded()
                    ? new SessionResolution(session, null)
                    : new SessionResolution(null, result.getProblem());
        }
    }

    private SessionResolution refreshAfterUnauthorized(String rejectedAccessToken) throws IOException {
        synchronized (sessionGate) {
            throwIfClosed();
            SessionState current = session;
            if (current != null && !current.accessToken.equals(rejectedAccessToken)) {
                return new SessionResolution(current, null);
            }
            RefreshTokenCredential credential = refreshTokenStore.load();
            if (credential == null) return new SessionResolution(null, SESSION_MISSING);
            PokeFolioAuthenticationResult result = refreshCore(credential);
            return result.isSucceeded()
                    ? new SessionResolution(session, null)
                    : new SessionResolution(null, result.getProblem());
        }
    }

    private void activateSession(PokeFolioApiPayloads.AuthEnvelope envelope) throws IOException {
        try {
            refreshTokenStore.save(new RefreshTokenCredential(
                    envelope.device.id,
                    envelope.refreshToken));
        } catch (IOException | RuntimeException saveError) {
            session = null;
            try {
                refreshTokenStore.delete();
            } catch (IOException | RuntimeException deleteError) {
                saveError.addSuppressed(deleteError);
            }
            throw new IOException("Rotated refresh token could not be persisted.", saveError);
        }
        session = new SessionState(
                envelope.userId,
                envelope.accessToken,
                envelope.accessTokenExpiresAt,
                new PokeFolioDevice(
                        envelope.device.id,
                        envelope.device.name,
                        envelope.device.platform,
                        envelope.device.createdAt,
                        envelope.device.lastSeenAt));
    }

    private void invalidateLocalSession() throws IOException {
        session = null;
        refreshTokenStore.delete();
    }

    private static PokeFolioSession toPublicSession(SessionState value) {
        return value == null
                ? null
                : new PokeFolioSession(value.userId, value.device, value.accessTokenExpiresAt);
    }

    private void throwIfClosed() throws IOException {
        if (closed) throw new IOException("PokeFolio API client is closed.");
    }

    @Override
    public void close() throws IOException {
        synchronized (sessionGate) {
            if (closed) return;
            closed = true;
            session = null;
            transport.close();
        }
    }

    private static void clear(byte[] value) {
        if (value != null) Arrays.fill(value, (byte) 0);
    }

    private static final class SessionState {
        final java.util.UUID userId;
        final String accessToken;
        final java.time.Instant accessTokenExpiresAt;
        final PokeFolioDevice device;

        SessionState(
                java.util.UUID userId,
                String accessToken,
                java.time.Instant accessTokenExpiresAt,
                PokeFolioDevice device
        ) {
            this.userId = userId;
            this.accessToken = accessToken;
            this.accessTokenExpiresAt = accessTokenExpiresAt;
            this.device = device;
        }
    }

    private static final class SessionResolution {
        final SessionState session;
        final PokeFolioApiProblem problem;

        SessionResolution(SessionState session, PokeFolioApiProblem problem) {
            this.session = session;
            this.problem = problem;
        }
    }
}
